XMonad/Actions/RepeatAction.hs

```diff
@@ -0,0 +1,60 @@
+-----------------------------------------------------------------------------
+-- |
+-- Module      :  XMonad.Actions.RepeatAction
+-- Description :  Repeat the last performed action.
+-- Copyright   :  (c) 2022 Martin Kozlovsky
+-- License     :  BSD3-style (see LICENSE)
+--
+-- Maintainer  :  <kozlovsky.m7@gmail.com>
+-- Stability   :  unstable
+-- Portability :  not portable
+--
+-- Ability to repeat the last action.
+--
+-----------------------------------------------------------------------------
+
+module XMonad.Actions.RepeatAction (
+        -- * Usage
+        -- $usage
+          XMonad.Actions.RepeatAction.rememberAction
+        , XMonad.Actions.RepeatAction.rememberActions
+        , XMonad.Actions.RepeatAction.repeatLast
+) where
+
+import XMonad
+
+import qualified XMonad.Util.ExtensibleState as XS
+
+-- $usage
+--
+-- You can use this module with the following in your @xmonad.hs@:
+--
+-- > import XMonad.Actions.RepeatAction
+--
+-- Then use the 'rememberActions' on the list of your keybindings and
+-- add a dedicated key to run the last performed action:
+--
+-- > myKeys = ((modm, xK_period), repeatLast) : rememberActions [((modm, xK_c), kill), ...]
+--
+-- / Be careful not to include the key binding for the repetition in the remembered actions! /
```

Comment:
Indeed, that would be quite bad.  Perhaps then a better interface would be to not expose `repeatLast` individually, but only in combination with `rememberActions`?  Something like `rememberActions :: (KeyMask, KeySym) -> [((KeyMask, KeySym), X ())] -> XConfig a -> XConfig a`?  This could then be used as a combinator.  Might also be worth having a `rememberActionsP :: String -> [(String, X ())] -> XConfig l -> XConfig l`

Reply:
I'm a little confused about the type signature and how the function could be easily used, especially if one's key bindigs are a function of `XConfig`. So far I at least changed the interface to safer `rememberActions :: a -> [(a, X ())] -> [(a, X ())]` (which can be used nicely in infix form). I stil expose `repeatLast` as one might want to use it outside of theirs key bindings for some reason.

Reply:
So, for example, if one has a config like

``` haskell
main :: IO ()
main = xmonad
     . …
     $ def
       { …
       }
      `additionalKeysP` myKeys
```

Then with what I was thinking of (`rememberActionsP :: String -> [(String, X ())] -> XConfig l -> XConfig l`) this would just be transformed to

``` haskell
main :: IO ()
main = xmonad
     . …
     . rememberActionsP "M-." myKeys
     $ def
       { …
       }
```

This looks quite ergonomic to me.  Even if one doesn't want all keys to be remembered, this would then just incur adding `additionalKeysP` back to the bottom of main with some `myKeys'` instead.

The current implementation would give

``` haskell
main :: IO ()
main = xmonad
     . …
     $ def
       { …
       }
      `additionalKeysP` (rememberActions "M-." myKeys)
```

which isn't bad either, especially since one can do

``` haskell
      `additionalKeysP` myKeys'
      `additionalKeysP` (rememberActions "M-." myKeys'')
```


If one's keybindings are a function `XConfig -> Keybindings` then even today one has to directly override the `keys` field; in this case you could just write something like `keys = rememberActions "M-." . myKeys` (in your version; for the other `additionalKeys`-like `rememberActions` this would be a little bit more complicated, depending on the exact implementation) and it should be fine

Reply:
I see. So can I mark this conversation as resolved since the current implementetion seems sufficient enough?

Reply:
I don't have any strong preferences here; if you like the current implementation I'd say let's go with that!